### PR TITLE
implementation

### PR DESCRIPTION
**Pull request by JaSei**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,23 @@
...
+
+[![Release](https://img.shields.io/github/release/avast/hashutil-go.svg?style=flat-square)](https://github.com/avast/hashutil-go/releases/latest)
+[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
+[![Travis](https://img.shields.io/travis/avast/hashutil-go.svg?style=flat-square)](https://travis-ci.org/avast/hashutil-go)
+[![Go Report Card](https://goreportcard.com/badge/github.com/avast/hashutil-go?style=flat-square)](https://goreportcard.com/report/github.com/avast/hashutil-go)
+[![GoDoc](https://godoc.org/github.com/avast/hashutil-go?status.svg&style=flat-square)](http://godoc.org/github.com/avast/hashutil-go)
+[![codecov.io](https://codecov.io/github/avast/hashutil-go/coverage.svg?branch=master)](https://codecov.io/github/avast/hashutil-go?branch=master)
+[![Sourcegraph](https://sourcegraph.com/github.com/avast/hashutil-go/-/badge.svg)](https://sourcegraph.com/github.com/avast/hashutil-go?badge)
+
+hashutil is collection of common hash (crypto) types
```
Comment (alenkacz) [inline README.md]: is a collection of common hash (crypto) types. It also contains some helper functions for hash manipulation.

... or something like that :)

File: md5.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+			return nil, errors.New("Md5 must contains only 0-FF bytes")
+		}
+	}
+
+	return bytes, nil
+}
+
+// EmptyMd5 return type Md5 of empty string
+func EmptyMd5() Md5 {
+	return md5.New().Sum(nil)
```
Comment (alenkacz) [inline md5.go]: how come is empty string a valid md5? I would expect that it should not allow to create invalid md5

File: md5_test.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
+package hashutil
+
+import (
+	"crypto/md5"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+const emptyMd5 = "d41d8cd98f00b204e9800998ecf8427e"
```
Comment (alenkacz) [inline md5_test.go]: don't understand why is this called empty, maybe testMd5?

File: sha256.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
+/*
```
Comment (alenkacz) [inline sha256.go]: why is the synopsis in this file?

File: md5.go (hunk 2 above)
Comment (alenkacz) [reply md5.go]: the same goes for sha256 and sha1 types...

**Review by alenkacz — CHANGES_REQUESTED** [review]

File: md5.go (hunk 2 above)
Comment (JaSei) [reply md5.go]: it's like `echo -n | md5sum` it's hash of `""` string (empty/zero length)
https://en.wikipedia.org/wiki/MD5

File: md5_test.go (hunk 3 above)
Comment (JaSei) [reply md5_test.go]: its hex representation of `zero-length hash`

File: sha256.go (hunk 4 above)
Comment (JaSei) [reply sha256.go]: godoc concat all files together and doc header from some file use as header... is possible to extract this to some special file like `doc.go`. maybe it's more comprehensible? (like https://golang.org/src/fmt/doc.go)

File: README.md (hunk 1 above)
Comment (JaSei) [reply README.md]: thanks :+1:

File: md5.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+type Md5 []byte
+
+var badMd5Length = fmt.Errorf("Md5 must have %d bytes length", md5.Size)
+
+// convert type Md5 to string representation (upper case)
+func (hash Md5) String() string {
+	return hex.EncodeToString(hash)
+}
+
+// StringToMd5 valid and convert string to Md5 type
```
Comment (petr-k) [inline md5.go]: "...validate and convert a string..."

File: md5.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+// BytesToMd5 valid and convert []bytes to Md5 type
+func BytesToMd5(bytes []byte) (Md5, error) {
+	if len(bytes) != md5.Size {
+		return nil, badMd5Length
+	}
+
+	for _, b := range bytes {
+		if b > 0xff {
```
Comment (petr-k) [inline md5.go]: What's the point of this check? Can a byte be greater than `0xff`?

File: sha1.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"errors"
+	"fmt"
+)
+
+type Sha1 []byte
+
+var badSha1Length = fmt.Errorf("Sha1 must have %d bytes length", sha1.Size)
```
Comment (petr-k) [inline sha1.go]: "... must have a length of N bytes"

File: md5.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+			return nil, errors.New("Md5 must contains only 0-FF bytes")
```
Comment (petr-k) [inline md5.go]: "must contain"

File: md5.go (hunk 6 above)
Comment (JaSei) [reply md5.go]: no, this check don't sense - fixed in f1fc7353400748825ffd472fe534381a26871cdf

File: md5.go (hunk 2 above)
Comment (petr-k) [reply md5.go]: Yet there is a test assertion here guarding against an empty string:
https://github.com/avast/hashutil-go/pull/1/files#diff-8928a00e3b4e3beedab48ebb046e7da2R25

I'm not sure why a value for an empty string is explicitly valuable here..

File: sha256_test.go [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+	s2, err := StringToSha256("E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855")
+	assert.NoError(t, err, "Conversion string to Sha256 without errors")
+
+	assert.Equal(t, s1, s2, "Upper and lower of same Sha256 are equal")
+
+	h := sha256.New()
+	assert.Equal(t, s1, (Sha256)(h.Sum(nil)), "Empty hash calculated by crypto library are equal to hash from string")
+
+	assert.Equal(t, s1.String(), emptySha256, "Converse Sha256 to strings")
```
Comment (petr-k) [inline sha256_test.go]: "Convert Sha256 to..."

File: sha256_test.go [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+	s1, err := StringToSha256(emptySha256)
+	assert.NoError(t, err, "Conversion string to Sha256 without errors")
+
+	s2, err := StringToSha256("E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855")
+	assert.NoError(t, err, "Conversion string to Sha256 without errors")
+
+	assert.Equal(t, s1, s2, "Upper and lower of same Sha256 are equal")
+
+	h := sha256.New()
+	assert.Equal(t, s1, (Sha256)(h.Sum(nil)), "Empty hash calculated by crypto library are equal to hash from string")
```
Comment (petr-k) [inline sha256_test.go]: "An empty hash... is equal to..."

File: sha256_test.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+const emptySha256 = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
+
+func TestSha256(t *testing.T) {
+	s1, err := StringToSha256(emptySha256)
+	assert.NoError(t, err, "Conversion string to Sha256 without errors")
```
Comment (petr-k) [inline sha256_test.go]: "Convert string to... without errors"

File: md5.go (hunk 6 above)
Comment (JaSei) [reply md5.go]: Sorry, I forgot push this change... Now is fixed..

File: sha1.go (hunk 7 above)
Comment (JaSei) [reply sha1.go]: thanks

**Pull request merged**